### PR TITLE
Simplify MetaModel input handling

### LINKED ISSUE #1624: MetaModel should have a dictionary interface only
The meta model class here 
https://github.com/NNPDF/nnpdf/blob/94b80c35618bd55ae463b5806bf2b5886f6ecad6/n3fit/src/n3fit/backends/keras_backend/MetaModel.py#L83
is supposed to be a thin wrapper over a Keras model, where one of the differences is that it works around the peculiarity of keras where one has to bend over backwards to pass constant tensors to layers (i.e. https://github.com/keras-team/keras/issues/11912).

However the way in which it does it is a bit convoluted. One can pass a dictionary like for a keras model, but then the information of the keys is discarded, 

https://github.com/NNPDF/nnpdf/blob/94b80c35618bd55ae463b5806bf2b5886f6ecad6/n3fit/src/n3fit/backends/keras_backend/MetaModel.py#L112

and new keys are added, with some complicated logic

https://github.com/NNPDF/nnpdf/blob/94b80c35618bd55ae463b5806bf2b5886f6ecad6/n3fit/src/n3fit/backends/keras_backend/MetaModel.py#L142

which happens on depend on the keras logic not doing things like

```
In [1]: import tensorflow as tf

In [2]: name = "I really want this name"

In [3]: print(tf.keras.Input(tensor=tf.constant([1,2,3], name=name), name=name).name)

None
```

and then more cumbersome logic to recall the proper inputs, like e.g.

https://github.com/NNPDF/nnpdf/blob/94b80c35618bd55ae463b5806bf2b5886f6ecad6/n3fit/src/n3fit/backends/keras_backend/MetaModel.py#L56

which can get tricky both for the class and the caller.

I think a better way would be to mandate that the input argument is only allowed to be a dict, and similarly allow only dict predictions and similar. Then the keys are the users responsibility, and we can easily produce better errors if some input is not given, rather than some hard to decipher message down the line.

### PR DESCRIPTION
**Pull request by Zaharid**
Mandate that all input is passed as dictionaries. This has the advantage of doing away with a lot of code, having an obvious way to derive the name for each input (the dictionary key) and allowing for better error messages.

Closes #1624

### CODE REVIEW

**Comment by Zaharid** [comment]
@scarlehoff if you agree with this in principle I'll work on tests and so on.

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 1]
```diff
@@ -103,71 +77,54 @@ class MetaModel(Model):
...
-            input_dict[name] = input_tensor
-            try:
-                x_in[name] = op.numpy_to_tensor(input_tensor.tensor_content)
-                tensors_in[name] = input_tensor
-            except AttributeError:
-                x_in[name] = None
-                tensors_in[name] = None
-
-        super().__init__(input_dict, output_list, **kwargs)
+        for k, v in input_tensors.items():
```
Comment (scarlehoff) [inline n3fit/src/n3fit/backends/keras_backend/MetaModel.py:91]: Please add some comments here

**Review by scarlehoff — COMMENTED** [review]
I'm happy with the change _in principle_.
But the convoluted way of doing this was because of a change at some point on how Tensorflow was handling things so you need to make sure it works in all possible scenarios that we have (in practice, at least with all the runcards in the example folder).

i.e., parallelization over replicas, multiples replicas, feature scaling, with/without sum rules, etc.

I'm actually not even sure whether the dictionary is currently the only option used so that most of the code that was there is currently unused.

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 2]
```diff
@@ -103,71 +77,54 @@ class MetaModel(Model):
...
         """
-        if isinstance(extra_input, dict) and self._scaler is not None:
+        if extra_input is None:
+            if self.required_slots:
+                raise ValueError(
+                    f"The following inputs must be provided: {self.required_slots}"
+                )
+            return self.x_in
+
+        if diff := (self.required_slots - extra_input.keys()):
```
[the remark this answers is not in the document]
Comment (Zaharid) [reply n3fit/src/n3fit/backends/keras_backend/MetaModel.py:121]: Don't remember the context and I do dislike it overall. Maybe something to do with it not working for old python? Or requiring too many parentheses?). I don't mind either way, but looks okay to me in here.

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 3]
```diff
@@ -187,10 +144,10 @@ def perform_fit(self, x=None, y=None, epochs=1, **kwargs):
             loss_dict: dict
                 a dictionary with all partial losses of the model
         """
-        x = self._parse_input(x)
+        params = self._parse_input(x)
         if y is None:
             y = self.target_tensors
-        history = super().fit(x=x, y=y, epochs=epochs, **kwargs)
+        history = super().fit(x=params, y=y, epochs=epochs, **kwargs)
```
[the remark this answers is not in the document]
Comment (scarlehoff) [reply n3fit/src/n3fit/backends/keras_backend/MetaModel.py:150]: parsed x or whatever would also be fine, they are just not parameters so could get confused.

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 4]
```diff
@@ -53,121 +53,87 @@ def _default_loss(y_true, y_pred): # pylint: disable=unused-argument
...
-                x_in[name] = op.numpy_to_tensor(input_tensor.tensor_content)
-                tensors_in[name] = input_tensor
-            except AttributeError:
-                x_in[name] = None
-                tensors_in[name] = None
-
-        super().__init__(input_dict, output_list, **kwargs)
+        # Go over the inputs. If we can deduce a constant value, either because
+        # it is set in input_values or because it has a tensor_content, we
+        # store it. Otherwise we mar the input as required when making
```
Comment (scarlehoff) [inline n3fit/src/n3fit/backends/keras_backend/MetaModel.py:98]:
```suggestion
        # store it. Otherwise we mark the input as required when making
```

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 5]
```diff
@@ -53,121 +53,87 @@ def _default_loss(y_true, y_pred): # pylint: disable=unused-argument
...
-        super().__init__(input_dict, output_list, **kwargs)
+        # Go over the inputs. If we can deduce a constant value, either because
+        # it is set in input_values or because it has a tensor_content, we
+        # store it. Otherwise we mark the input as required when making
+        # predictions.
+        for k, v in input_tensors.items():
+            if k in input_values:
+                x_in[k] = input_values[k]
+            elif hasattr(v, "tensor_content"):
+                x_in[k] = v.tensor_content
```
Comment (scarlehoff) [inline n3fit/src/n3fit/backends/keras_backend/MetaModel.py:104]: You should keep the `op.numpy_to_tensor` here.

The reason for that is having one single place where to cast the numpy array to other types (if we want to do `double` instead of `float` or whatever).

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 6]
```diff
@@ -53,121 +53,87 @@ def _default_loss(y_true, y_pred): # pylint: disable=unused-argument
...
-        if isinstance(extra_input, dict) and self._scaler is not None:
+        if extra_input is None:
+            if self.required_slots:
+                raise ValueError(
+                    f"The following inputs must be provided: {self.required_slots}"
+                )
+            return self.x_in
+
+        if diff := (self.required_slots - extra_input.keys()):
+            raise ValueError(f"The following inputs must be provided {diff}")
```
Comment (scarlehoff) [inline n3fit/src/n3fit/backends/keras_backend/MetaModel.py:131]: A few comments here:

- `extra_input` could enter as a list (or as whatever without `keys`). Since we are checking for mistakes, this should be tested for (and the error message should also include the keys that are required).
- The `diff` should also check whether there are `extra_inputs` being provided that is not required.

**Review by scarlehoff — COMMENTED** [review]
You will also need to modify penalties https://github.com/NNPDF/nnpdf/blob/94b80c35618bd55ae463b5806bf2b5886f6ecad6/n3fit/src/n3fit/hyper_optimization/penalties.py#L58

For the rest I think it is ok.

However, now one cannot simply call the model like one would do for another model with some arrays and get some results but have to actually know the keys, which I don't like.

But I guess the simplification is worth it and this ^ is a developer use case (which used to be only me). 
But please, could you add more stuff to the MetaModel docstring explaining this? With an example with its ```>>>``` and everything.

**Comment by Zaharid** [comment]
> However, now one cannot simply call the model like one would do for another model with some arrays and get some results but have to actually know the keys, which I don't like.

I don't see a way to do this, have the input tracking behavior, and have a reasonable interface.

**Comment by scarlehoff** [comment]
If you force-push every commit it's very difficult to keep track of what is changing.

**Comment by Zaharid** [comment]
Should be everything that you asked and none of what yo didn't ask :P.

**Comment by scarlehoff** [comment]
> Should be everything that you asked and none of what yo didn't ask :P.

Is this ready for review then?

**Comment by Zaharid** [comment]
Yes please.

**Review by scarlehoff — APPROVED** [review]
lgtm

File: n3fit/src/n3fit/backends/keras_backend/MetaModel.py [hunk 7]
```diff
@@ -53,121 +53,98 @@ def _default_loss(y_true, y_pred): # pylint: disable=unused-argument
...
-                tensors_in[name] = None
-
-        super().__init__(input_dict, output_list, **kwargs)
+        # Go over the inputs. If we can deduce a constant value, either because
+        # it is set in input_values or because it has a tensor_content, we
+        # store it. Otherwise we mark the input as required when making
+        # predictions.
+        for k, v in input_tensors.items():
+            if k in input_values:
+                x_in[k] = input_values[k]
```
[the remark this answers is not in the document]
Comment (Zaharid) [reply n3fit/src/n3fit/backends/keras_backend/MetaModel.py:106]: I would not do this: Would prefer it to be to the caller, e.g. if creating a keras Input manually or getting it from somewhere.

**Comment by scarlehoff** [comment]
Stop with the force pushes  :___
We can squash and merge if you want to have everything well organized.
